util/gpool/spmc/spmspool.go

```diff
@@ -0,0 +1,379 @@
+type Pool[T any, U any, C any, CT any, TF pooltask.Context[CT]] struct {
+	gpool.BasePool
+	workerCache   sync.Pool
+	workers       *loopQueue[T, U, C, CT, TF]
+	lock          sync.Locker
+	cond          *sync.Cond
+	taskCh        chan *pooltask.TaskBox[T, U, C, CT, TF]
+	options       *Options
+	stopCh        chan struct{}
+	consumerFunc  func(T, C, CT) U
+	capacity      atomic.Int32
+	running       atomic.Int32
+	state         atomic.Int32
+	waiting       atomic.Int32
+	heartbeatDone atomic.Bool
+}
+
+// NewSPMCPool create a single producer, multiple consumer goroutine pool.
+func NewSPMCPool[T any, U any, C any, CT any, TF pooltask.Context[CT]](name string, size int32, options ...Option) (*Pool[T, U, C, CT, TF], error) {
+	opts := loadOptions(options...)
+	if expiry := opts.ExpiryDuration; expiry <= 0 {
+		opts.ExpiryDuration = gpool.DefaultCleanIntervalTime
+	}
+	result := &Pool[T, U, C, CT, TF]{
+		BasePool: gpool.NewBasePool(),
+		taskCh:   make(chan *pooltask.TaskBox[T, U, C, CT, TF], 128),
+		stopCh:   make(chan struct{}),
+		lock:     gpool.NewSpinLock(),
+		options:  opts,
+	}
+	result.SetName(name)
+	result.state.Store(int32(gpool.OPENED))
+	result.heartbeatDone.Store(false)
+	result.workerCache.New = func() interface{} {
+		return &goWorker[T, U, C, CT, TF]{
+			pool: result,
+			exit: make(chan struct{}),
+		}
+	}
+	result.capacity.Add(size)
+	result.workers = newWorkerLoopQueue[T, U, C, CT, TF](int(size))
+	result.cond = sync.NewCond(result.lock)
+	// Start a goroutine to clean up expired workers periodically.
+	go result.purgePeriodically()
+	return result, nil
+}
+
+// purgePeriodically clears expired workers periodically which runs in an individual goroutine, as a scavenger.
+func (p *Pool[T, U, C, CT, TF]) purgePeriodically() {
+	heartbeat := time.NewTicker(p.options.ExpiryDuration)
+	defer func() {
+		heartbeat.Stop()
+		p.heartbeatDone.Store(true)
+	}()
+	for {
+		select {
+		case <-heartbeat.C:
+		case <-p.stopCh:
+			return
+		}
+
+		if p.IsClosed() {
+			break
+		}
+
+		p.lock.Lock()
+		expiredWorkers := p.workers.retrieveExpiry(p.options.ExpiryDuration)
+		p.lock.Unlock()
+
+		// Notify obsolete workers to stop.
+		// This notification must be outside the p.lock, since w.task
+		// may be blocking and may consume a lot of time if many workers
+		// are located on non-local CPUs.
+		for i := range expiredWorkers {
+			expiredWorkers[i].taskBoxCh <- nil
+			expiredWorkers[i] = nil
+		}
+
+		// There might be a situation where all workers have been cleaned up(no worker is running),
+		// or another case where the pool capacity has been Tuned up,
+		// while some invokers still get stuck in "p.cond.Wait()",
+		// then it ought to wake all those invokers.
+		if p.Running() == 0 || (p.Waiting() > 0 && p.Free() > 0) {
+			p.cond.Broadcast()
+		}
+	}
+}
+
+// Tune changes the capacity of this pool, note that it is noneffective to the infinite or pre-allocation pool.
+func (p *Pool[T, U, C, CT, TF]) Tune(size int) {
```

Comment:
Could you give a explain about what did T, U, C, CT, TF represent?

Reply:
add some explanatory comments at the Pool.